README.md
# Buildit website

This is the source code for the Buildit website: [https://buildit.wiprodigital.com/](https://buildit.wiprodigital.com/)

To provide feedback, raise bugs or request features, please use our [Buildit website feedback form](https://docs.google.com/forms/d/e/1FAIpQLScOeoF7XfYYs3R8b9K_zvSY0oVS92hAV82FXUOyy8JnDz2lNg/viewform).

If you want to get involved in the development of this website, then please see our [contribution guidelines](./CONTRIBUTING.md).

## Requirements

Node v8.10.0 or higher is required.

If you use [NVM](https://github.com/creationix/nvm) for managing Node, you can just issue the following command to install the right node version

    $ nvm install

and before running any other npm command, run

    $ nvm use 


## Build and development

There are three ways to build and run the website:

| Script | Description |
|:--|:--|
| `npm run build` | Generates a minified local build |
| `npm run start` | Generates a minified local build and a local server instance |
| `npm run dev` | Generates a non minified local build and a local server instance |

Local hosting is at http://localhost:8080 and a browser window should open automatically.

Distribution goes into `./dist`.

Refer to the [build process](./docs/build-process.md) document for further details on all the tasks the build performs.

### Committing

All commits should be made with [Commitizen](https://github.com/commitizen/cz-cli). To make your commits with it 
simply run the following command.
 
`npm run commit`
 
Failure to format your commits properly will result in a rejected PR and/or failed cloud builds.

### Environments

Some of the files produced by the build, for example `sitemap.xml`, need to contain the website's absolute URL. Others, such as `robots.txt`, need to have different contents depending on where the build will be deployed to (so that we can prevent search engines indexing our staging environments).

To facilitate this, multiple environments can be defined in the `envs` section of [`config.json`](./config.json). Both the `npm start` and `npm run-script build` commands support an additional `--env` argument which takes the name of the desired environment for that build as its parameter (note that you need to proceed it with `--` so that `npm` passes that argument through to the underlying build script). For example:

    $ npm run-script build -- --env production

Where `production` corresponds to the key of the desired environment defined in `config.json`:

```js
{
  // ...
  "envs": {

    "production": {

      // The absolute URL that this site is deployed to
      "url": "https://buildit.foo.bar",

      // Whether or not search engine bots should be
      // prevented from indexing this site
      "excludeRobots": true

    },

    "other-env": {
      // ...
    },
    // ...
  },
  // ...
}
```

If no `--env` argument is provided to the builds, then the first environment defined in `config.json` will be used.

## Branches

For complete info see our [branching strategy docs](./docs/branching-strategy.md). The basic premise is that `master` is always correct and deployed, `develop` is used for creating feature branches and to then merge in to `master`. `next` is "special" and is used to build and test the _next_ big version of the app.

## Deployment

The app will auto-deploy to [production](https://buildit.wiprodigital.com/) using Travis with every passing merge to `master`.

It will also auto-deploy to the [`next` environment](www-next.buildit.digital/) with every passing merge to the `next` branch.

## More information

### General
* [Contribution guidelines](./CONTRIBUTING.md)
* [Coding standards](./docs/coding-standards.md)
* [Branching strategy](./docs/branching-strategy.md)
* [Testing](./docs/tests.md)
* [Build process](./docs/build-process.md)
* [Build and deploy pipeline](./docs/build-deploy-pipeline.md)

### Design
* [Flourish design elements](./docs/flourishes.md) to use, extend or create the flourish component

### Website technologies
Behind the scenes the website is using the following technologies, so be sure you know what you're doing before starting to change anything.

* [Metalsmith](http://metalsmith.io) for the core of the website,
* [HandlebarJS](https://handlebarsjs.com/) as templating engine,
* [`SCSS`](http://sass-lang.com/), with the addition of
  * [autoprefixer](https://github.com/postcss/autoprefixer)
  * [csso](https://github.com/css/csso)
  * [eyeglass](https://github.com/sass-eyeglass/eyeglass)
* [Cypress](./docs/tests.md), for end-to-end testing
* [Pa11y-CI](./docs/tests.md), for automated accessibility

### Decisions
Decisions linked to implementation details, have not been catalogued. Many of those decisions are now lost.  
Therefore, although late, we have decided to start logging all new decisions. These decisions will now be catalogued in the [Decision Log](./docs/DECISIONLOG.md)


CHANGELOG.md
# [4.8.0](https://github.com/buildit/buildit/compare/v4.7.0...v4.8.0) (2020-08-27)


### Features

* remove ex studios ([17635b1](https://github.com/buildit/buildit/commit/17635b1ece469e6a467f01c03ee607ad4844e359))

# [4.7.0](https://github.com/buildit/buildit/compare/v4.6.1...v4.7.0) (2020-08-05)


### Bug Fixes

* **header:** corrects logo position on narrow viewports ([8e14984](https://github.com/buildit/buildit/commit/8e1498491723ead1c7023ce7dc9e9686b1809b54))
* **header:** removes conflicting layout hack ([afff0cf](https://github.com/buildit/buildit/commit/afff0cf3fa85dc39a4b8a26ab259212bcb1afbf3))
* corrects buildit name in title ([5fb1aa9](https://github.com/buildit/buildit/commit/5fb1aa93e99cc73a3263eaa2901aea982d0e6c45))


### Features

* lowercases instances of "buildit" ([21ef486](https://github.com/buildit/buildit/commit/21ef48621828154d415f2e3dfe547b8572df5741))
* updates favicons to new logo ([c0dcc9d](https://github.com/buildit/buildit/commit/c0dcc9d0a402154664bb84446194da1ef01da39a))
* updates to new Gravity v2 release for new Buildit logo ([afad6e6](https://github.com/buildit/buildit/commit/afad6e69438cbbf3c164a79614ac06c6533c2477))
* updates wording on homepage ([7522f68](https://github.com/buildit/buildit/commit/7522f68071ef69ed7b4cb6e1f5516fac2ccc4be5))

## [4.6.1](https://github.com/buildit/buildit/compare/v4.6.0...v4.6.1) (2020-04-17)


### Bug Fixes

* solves issues when there are no open job positions ([8d413f5](https://github.com/buildit/buildit/commit/8d413f5e2ca49bc0bd6aee0a21044881d1d87301))

# [4.6.0](https://github.com/buildit/buildit/compare/v4.5.2...v4.6.0) (2020-02-06)


### Features

* **location:** Updated location of Warsaw's office ([98c4ed0](https://github.com/buildit/buildit/commit/98c4ed0c76d8430287d966e943f4f98223fcafb4))

## [4.5.2](https://github.com/buildit/buildit/compare/v4.5.1...v4.5.2) (2020-01-09)


### Bug Fixes

* fixes broken job posting links ([2ae62f0](https://github.com/buildit/buildit/commit/2ae62f065e09dc93b564825d32171d58561bd846))

## [4.5.1](https://github.com/buildit/buildit/compare/v4.5.0...v4.5.1) (2019-08-30)


### Bug Fixes

* fixes alignment of social media links and lists ([a494196](https://github.com/buildit/buildit/commit/a494196))

# [4.5.0](https://github.com/buildit/buildit/compare/v4.4.2...v4.5.0) (2019-04-23)


### Features

* dark mode ([aee7954](https://github.com/buildit/buildit/commit/aee7954))

## [4.4.2](https://github.com/buildit/buildit/compare/v4.4.1...v4.4.2) (2019-03-26)


### Bug Fixes

* pin eyeglass to 2.2.1 ([cca5a2b](https://github.com/buildit/buildit/commit/cca5a2b))
* **homepage:** fixes broken hamburger menu ([78003f8](https://github.com/buildit/buildit/commit/78003f8))
* **homepage:** fixes logo positioning in header ([e7b9a31](https://github.com/buildit/buildit/commit/e7b9a31))
* **locations:** fixes Plano studio title ([662da77](https://github.com/buildit/buildit/commit/662da77))
* **locations:** updates Denver studio address ([91c2236](https://github.com/buildit/buildit/commit/91c2236))

## [4.4.1](https://github.com/buildit/buildit/compare/v4.4.0...v4.4.1) (2019-03-25)


### Bug Fixes

* **locations:** fixes Plano studio title ([b01aed5](https://github.com/buildit/buildit/commit/b01aed5))
* **locations:** updates Denver studio address ([e17dcff](https://github.com/buildit/buildit/commit/e17dcff))
* pin eyeglass to 2.2.1 ([b628f8f](https://github.com/buildit/buildit/commit/b628f8f))

# [4.4.0](https://github.com/buildit/buildit/compare/v4.3.1...v4.4.0) (2019-03-18)


### Bug Fixes

* switched dependency due to issue with subdependency ([5b9d574](https://github.com/buildit/buildit/commit/5b9d574))


### Features

* adds Schema.org place meta-data to locations page ([5819024](https://github.com/buildit/buildit/commit/5819024))
* adds Schema.org RDFa data to job listings ([9062837](https://github.com/buildit/buildit/commit/9062837))
* adds Schema.org RDFa data to page header & footer ([4cf20f0](https://github.com/buildit/buildit/commit/4cf20f0))

## [4.3.1](https://github.com/buildit/buildit/compare/v4.3.0...v4.3.1) (2019-03-11)


### Bug Fixes

* **travis:** production was using the wrong release script ([4afd8c4](https://github.com/buildit/buildit/commit/4afd8c4))

# [4.3.0](https://github.com/buildit/buildit/compare/v4.2.0...v4.3.0) (2019-03-11)


### Bug Fixes

* **locations:** renames Dallas to Plano ([b315b79](https://github.com/buildit/buildit/commit/b315b79))


### Features

* **analytics:** switch to GTM and track smart recruiters links ([1cf2187](https://github.com/buildit/buildit/commit/1cf2187))
* **careers:** adds talent team email links ([0e14695](https://github.com/buildit/buildit/commit/0e14695))
* **commits:** support for commitizen formatted commits ([cfd4f21](https://github.com/buildit/buildit/commit/cfd4f21))
* **semantic-release:** added semantic now runs on `master` branch ([a4d8f57](https://github.com/buildit/buildit/commit/a4d8f57))


## [4.2.0](https://github.com/buildit/buildit/compare/v4.1.4...v4.2.0) (2019-01-15)
### Changed
- Upgraded to Gravity v0.11.0
- Upgraded dev dependencies
- Travis builds now use Node version specified in .nvmrc
- Enabled [Greenkeeper](https://greenkeeper.io/) to automatically update depencies
- `scrollreveal` is now included in our JS bundle, instead of being loaded via CDN
- Updated Warsaw studio address

### Fixed
- `scrollreveal` is no longer incorrectly listed as a (non-dev) dependency
- Removed unused dependencies


## [4.1.4](https://github.com/buildit/buildit/compare/v4.1.3...v4.1.4) (2018-10-23)
### Fixed
- Corrected capitalisation of "Buildit @ Wipro Digital" text


## [4.1.3](https://github.com/buildit/buildit/compare/v4.1.2...v4.1.3) (2018-09-24)
### Fixed
- Fixes a typo on the home page 

### Added
- Dallas office to locations page


## [4.1.2](https://github.com/buildit/buildit/compare/v4.1.1...v4.1.2) (2018-08-16)
### Added
- Warsaw office to locations page

### Fixed

logAggregator.sh
#!/bin/bash

function isDateValid {
    date_var=$1;

    if [[ $date_var =~ ^[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}$ ]]; then
        if date --date "$ymd" >/dev/null 2>&1; then
            echo "Date Valid! Proceeding...";
            return 0;
        else
            echo "Date is Invalid! Please fix";
            return 1;
        fi
    else
        echo "Date $date_var is not in the correct format: YYYY-mm-dd";
        return 1;
    fi
}

function getDateForAWS {
    date_var=$1
    if isDateValid $date_var; then
        echo "Date $date_var is not in the correct format: YYYY-mm-dd";
    fi
    
    if [[ $DATE =~ ^20[0-9]{2}-[0-1][0-12]-[0-3]{1,2}$ ]]; then
        echo "Passed 1st Validation";
        year=`echo $DATE | cut -d'-' -f1`;
        month=`echo $DATE | cut -d'-' -f2`;
        day=`echo $DATE | cut -d'-' -f3`;

        if [ "$month" == "1" ]; then
            month="01";
        elif [ "$month" == "2" ]; then
            month="02"
        elif [ "$month" == "3" ]; then
            month="03  "
        elif [ "$month" == "4" ]; then
            month="04"
        elif [ "$month" == "5" ]; then
            month="05"
        elif [ "$month" == "6" ]; then
            month="06"
        elif [ "$month" == "7" ]; then
            month="07"
        elif [ "$month" == "8" ]; then
            month="08"
        elif [ "$month" == "9" ]; then
            month="09"
        fi

        ymd=$year"-"$month"-"$day
        echo "ymd: "$ymd
        if date --date "$ymd" >/dev/null 2>&1; then
            echo "Date Valid! Proceeding...";
            return 0;
        else
            echo "Date is Invalid! Please fix";
            return 1;
        fi
    else
        echo "Date $DATE is not in the correct format"
        return 1
    fi
}

function run_log_pull {
    aws_date=$1;
    log_folder="aws_logs_$aws_date"
    mkdir -p "${log_folder}"
    echo "Starting Sync into $log_folder"
    echo "" aws s3 sync $2 "./${log_folder}" --delete --exclude \'*\' --include \'*$aws_date*\'
    aws s3 sync $2 "./${log_folder}" --delete --exclude '*' --include "\'*$aws_date*\'" --exclude 'FullLog_*.txt' 
}

function run_log_join {
    aws_date=$1;
    log_folder="aws_logs_$aws_date"
    mkdir -p "${log_folder}/combined_logs"
    cd "$log_folder";
    touch ./combined_logs/FullLog_${aws_date}.txt
    i=0
    for fle in ./*${aws_date}* ; do
        echo $(i=i+1);
        if [ -f $fle ]; then
            while IFS='' read -r line || [[ -n "$line" ]]; do
                echo "3: $line" 
                echo "$line" >> ./combined_logs/FullLog_${aws_date}.txt
            done < $fle
        fi
    done
    #    done < <(find . -maxdepth 1 -iname "*$1*" )
}

echo "============================== AWS Log Sync Start ==========================================="
echo "Info:"
echo "This script assumes you have aws tools installed and configured correctly (inc. access keys)."
echo "If you do not, you must download and configure aws CLI."
echo " "
echo "============================================================================================"
echo "Please bear in mind, if you are investigating for issues around midnight, the majority of "
echo "these requests may be in the logs after the day you are currently investigating, due to AWS."
echo "Note:"
echo "The script can take up to 2 minutes of what seems like idle time, this is AWS filtering out "
echo "all the logs to only download the specified ones. S.A.F."
echo " "
echo "============================================================================================"
echo " "
read -p "To start, please enter the date in format: YYYY-MM-dd : "  answer;
# DATE=`date +%Y-%m-%d` # current date `date '+%Y-%m-%d %H:%M:%S'`

declare s3_bucket=$1;
declare aws_date;
if [[ -z ${s3_bucket} ]]; then
    echo "User has NOT overridden S3 bucket set"
    s3_bucket="s3://buildit.wiprodigital.com-logs/logs/buildit.wiprodigital.com/"
fi
if [[ $answer =~ ^20[0-9]{2}-[0-1]?[0-9]-[0-3]?[0-9]$ ]]; then 
    run_log_pull $answer $s3_bucket
    run_log_join $answer;
else
    echo "Enter Date in format YYYY-mm-dd, please.";
fi

echo "============================ AWS Log Sync Complete ========================================="